dbt/context/common.py

```diff
@@ -414,27 +408,75 @@ def generate(model, project_cfg, flat_graph, provider=None):
+    if model.resource_type == NodeType.Operation:
         this = db_wrapper.adapter.Relation.create(
                 schema=target['schema'],
-                identifier=model['name']
+                identifier=model.name
         )
     else:
-        this = get_this_relation(db_wrapper, project_cfg, profile, model)
+        this = get_this_relation(db_wrapper, project_cfg, profile, model_dict)
 
     context["this"] = this
+    return context
+
+
+def modify_generated_context(context, model, model_dict, project_cfg,
+                             manifest):
+    cli_var_overrides = project_cfg.get('cli_vars', {})
 
     context = _add_tracking(context)
     context = _add_validation(context)
     context = _add_sql_handlers(context)
 
     # we make a copy of the context for each of these ^^
 
-    context = _add_macros(context, model, flat_graph)
+    context = _add_macros(context, model, manifest)
 
-    context["write"] = write(model, project_cfg.get('target-path'), 'run')
-    context["render"] = render(context, model)
+    context["write"] = write(model_dict, project_cfg.get('target-path'), 'run')
+    context["render"] = render(context, model_dict)
     context["var"] = Var(model, context=context, overrides=cli_var_overrides)
     context['context'] = context
 
     return context
+
+
+def generate_operation_macro(model, project_cfg, manifest, provider):
+    """This is an ugly hack to support the fact that the `docs generate`
+    operation ends up in here, and macros are not nodes.
+    """
+    model_dict = model.serialize()
+    context = generate_base(model, model_dict, project_cfg, manifest,
+                            None, provider)
+
+    return modify_generated_context(context, model, model_dict, project_cfg,
+                                    manifest)
+
+
+def generate_model(model, project_cfg, manifest, source_config, provider):
+    model_dict = model.to_dict()
+    context = generate_base(model, model_dict, project_cfg, manifest,
+                            source_config, provider)
+    # overwrite schema if we have it, and hooks + sql
+    context.update({
+        'schema': model.get('schema', context['schema']),
+        'pre_hooks': model.config.get('pre-hook'),
+        'post_hooks': model.config.get('post-hook'),
+        'sql': model.get('injected_sql'),
+    })
+
+    return modify_generated_context(context, model, model_dict, project_cfg,
+                                    manifest)
+
+
+def generate(model, project_cfg, manifest, source_config=None, provider=None):
+    """
+    Not meant to be called directly. Call with either:
+        dbt.context.parser.generate
+    or
+        dbt.context.runtime.generate
+    """
+    if isinstance(model, ParsedMacro):
+        return generate_operation_macro(model, project_cfg, manifest, provider)
+    else:
+        return generate_model(model, project_cfg, manifest, source_config,
+                              provider)
```

Comment:
which part of this are you unhappy with?

Reply:
I don't like how these two functions are implemented, where I do a setup that builds a partial context, and then some custom stuff, and then a finalization step. It's basically doing inheritance/class-based work but with functions.

I think the obvious change to make is to turn context generation into a class-based thing, where these both inherit from some common base class that implements the first/last and perform a common method that implements them. That seemed like a lot of high-risk work and I think context generation/rendering is going to need some more attention in the future anyway, so I left it like this to avoid locking us into anything.